xscen/ensembles.py

```diff
@@ -62,38 +73,184 @@ def ensemble_stats(
+        if (
+            weighted is not None
+            and "weights" in inspect.getfullargspec(getattr(ensembles, stat))[0]
+        ):
+            stats_kwargs.setdefault("weights", weights)
+        if "ref" in stats_kwargs:
+            stats_kwargs["ref"] = ens.loc[stats_kwargs["ref"]]
+
+        if stat == "change_significance":
+            for v in ens.data_vars:
+                with xc.set_options(keep_attrs=True):
+                    ens_stats[f"{v}_change_frac"], ens_stats[f"{v}_pos_frac"] = getattr(
+                        ensembles, stat
+                    )(ens[v], **stats_kwargs)
+        else:
+            ens_stats = ens_stats.merge(getattr(ensembles, stat)(ens, **stats_kwargs))
+
+    # delete the realization coordinate if there
+    if "realization" in ens_stats:
+        ens_stats = ens_stats.drop_vars("realization")
 
     # delete attrs that are not common to all dataset
     if common_attrs_only:
-        for i in range(len(datasets)):
-            if isinstance(datasets[i], (str, Path)):
-                # if they exist remove attrs specific to create_ensemble
-                create_kwargs.pop("mf_flag", None)
-                create_kwargs.pop("resample_freq", None)
-                create_kwargs.pop("calendar", None)
-                create_kwargs.pop("preprocess", None)
-                ds = xr.open_dataset(datasets[i], **create_kwargs)
-            else:
-                ds = datasets[i]
-            attributes = ens_stats.attrs.copy()
-            for a_key, a_val in attributes.items():
-                if (
-                    (a_key not in ds.attrs)
-                    or (a_key in ["cat:date_start", "cat:date_end"])
-                    or (a_val != ds.attrs[a_key])
-                ):
-                    del ens_stats.attrs[a_key]
-        # create dataframe of catalogue attrs to generate new id
-        df = pd.DataFrame.from_dict(
-            {
-                key[4:]: [value]
-                for key, value in ens_stats.attrs.items()
-                if key.startswith("cat:")
-            }
+        # if they exist remove attrs specific to create_ensemble
+        create_kwargs.pop("mf_flag", None)
+        create_kwargs.pop("resample_freq", None)
+        create_kwargs.pop("calendar", None)
+        create_kwargs.pop("preprocess", None)
+
+        ens_stats = clean_up(
+            ds=ens_stats, common_attrs_only=datasets, xrkwargs=create_kwargs
         )
-        ens_stats.attrs["cat:id"] = generate_id(df)[0]
+
+        # generate new id
+        ens_stats.attrs["cat:id"] = generate_id(ens_stats).iloc[0]
 
     ens_stats.attrs["cat:processing_level"] = to_level
 
     return ens_stats
+
+
+def generate_weights(
+    ens: Union[xr.Dataset, xr.DataArray],
+    info: dict = None,
+    independence_level: str = "all",
+) -> xr.DataArray:
+    """
+    Uses realization attributes to automatically generate weights along the 'realization' dimension.
```

Comment:
I feel like this needs a bit more explanation as to how the weigths are generated.

Reply:
I tried to rework the Docstrings.

Reply:
much clearer, thank you!